[Feature] - Transactions

gvolpe opened the pull request:
closes #70 

- [x] Initial building blocks to support [transactions](https://redis.io/topics/transactions): `multi`, `exec`, `discard`, `watch` and `unwatch`.
- [x] High-level abstraction over transactions. ~Maybe `Resource[F, Transactional[F]]`~.
- [x] Write unit tests.
- [x] Write documentation.
- [x] Verify transactions work accordingly in a cluster setup (it does not for now, see #72)

Note that every single command needs to be forked (`.start`) or run in parallel since a `Future` cannot be lifted into an `F` until either an `EXEC` or `DISCARD` command is sent to the server.

See the [Lettuce Transactions API](https://github.com/lettuce-io/lettuce-core/wiki/Transactions).

/cc @Daenyth

gvolpe commented on modules/core/src/main/scala/com/github/gvolpe/fs2redis/connection/Fs2RedisConnection.scala:
```diff
@@ -19,24 +19,30 @@ import cats.effect.{ Concurrent, ContextShift, Sync }
-  override def close: F[Unit]                            = JRFuture.fromCompletableFuture(Sync[F].delay(conn.closeAsync())).void
+  override def async: F[RedisAsyncCommands[K, V]] = Sync[F].delay(conn.async())
+  override def clusterAsync: F[RedisClusterAsyncCommands[K, V]] =
+    Sync[F].raiseError(new Exception("Operation not supported"))
```
Can do that. Not sure if users would care about the Exception type though.

gvolpe commented on modules/effects/src/main/scala/com/github/gvolpe/fs2redis/transactions.scala:
```diff
@@ -0,0 +1,52 @@
+    def apply[F[_]: Log: Sync, K, V, A](cmd: RedisCommands[F, K, V])(fa: F[A]): F[A] =
+      resource(cmd).use(_ => fa)
+
+    def resource[F[_]: Log: Sync, K, V](cmd: RedisCommands[F, K, V]): Resource[F, Unit] =
```
I agree it's weird and it's because there's no resource. Maybe the API shouldn't expose a `resource` method? The `apply` should be more than enough and can be implemented using `bracketCase` directly. WDYT?

The only benefit of having it as a resource is that it can be composed but it might not be the way the user wants to use it.

gvolpe commented:
On the `start` need: when a transaction begins (`MULTI`), every command that is sent after register a callback semantically blocking (`CompletableFuture`) until it gets a result back. It will only get a result when either an `EXEC` or a `DISCARD` command is sent. 

Eg.

```scala
cmd.multi.bracket { _ => 
  cmd.set("k1", "foo") *> // this command will block forever and neither `exec` nor `discard` will ever be reached
    cmd.set("k2", "bar") *>
    cmd.exec // commit transaction
} (_ => cmd.discard) // rollback, discard transaction
```

Given the problem presented above, the commands need to be sent to the server with fire-and-forget semantics, thus I suggested using `start`. 

I do not know if this should be done in another way or not but based on the [Lettuce documentation](https://github.com/lettuce-io/lettuce-core/wiki/Transactions) this seems to be the way to do it. Actually, the examples in Java are very imperative so there's no sequence of commands, just side effects one after another.

I also thought about just `start`ing a set of commands but you could run into race conditions where `EXEC` runs before all the commands are sent to the server.

Daenyth commented:
What do you think about making a variation of the client implementation such that:
```scala
def resource[F[_]: Log: Sync, K, V](cmd: RedisCommands[F, K, V]): Resource[F, InTransactionCommands[F, K, V]] =
  Resource(... new TransactionCommands(cmd))
```

where `InTransactionCommands` provides RedisCommands (exclusing MULTI/EXEC/DISCARD) and encapsulates the required `Concurrent#start` behavior?

gvolpe commented:
Sounds like a good idea. I'll explore it these days and see what I can come up with. I would like to avoid code repetition (i.e. re-defining all the `get`, `set`, `del`, etc commands) and also have in mind the transaction support for cluster setups when designing this API.

gvolpe commented:
@Daenyth the more I look into it the less it makes to expose a "transactional" resource but I like the idea of hiding the `multi`, `exec` and `discard` commands. I think a simple `apply` will do:

```scala
commandsApi.use { cmd =>
  val tx = Transactional(cmd)

  val setters =
    List(
      cmd.set(key1, "foo"),
      cmd.set(key2, "bar")
    ).traverse_(_.start)

  val failedSetters =
    List(
      cmd.set(key1, "qwe"),
      cmd.set(key2, "asd")
    ).traverse(_.start) *> IO.raiseError(new Exception("boom"))
  
  tx.run(setters) *> tx.run(failedSetters)
}
```

> and encapsulates the required `Concurrent#start` behavior?

The reason why I don't think this is a good idea is because users might want to do things like:

```scala
val program = for {
  x <- cmd.get("foo")
  y <- cmd.get("bar")
  _ <- (x, y).tupled.fold(IO.unit) { case (a, b) => cmd.set("baz", a + b) }
} yield ()

tx.run(program)
```

They would be sequencing commands like this not knowing that this is not possible, even if you try to do it directly on Redis. The first command will be blocked until there's an `exec` or `discard` sent to the server. If we take care of calling `start` on `program` we might be creating a race condition where not all commands might be executed.

Now we could make this explicit and say that the users should provide a `List[IO[A]]` representing a list of independent commands and we can take care of `traverse(_.start)` but it's not much of a save. And the users might prefer to do something like this instead:

```scala
val setters =
  cmd.set(key1, "foo").start *> cmd.set(key2, "bar").start
```

In addition, I think a valid use case is collecting the results of every single transactional command by mutating a state. Eg:

```scala
Ref.of[IO, Map[String, String]](Map.empty).flatMap { ref =>
  cmd.get(key1).flatMap(x => ref.modify(_.updated(key1, x))).start *> 
    cmd.get(key2).flatMap(x => ref.modify(_.updated(key2, x))).start 
}

Daenyth commented:
Maybe the transaction commands need to return some kind of `TransactionIO[A]` monad, and we have a `TransactionIO` ~> `F` in the client? I'm not understanding why there inherently needs to be concurrency, it seems like it must be tied to some implementation detail.

gvolpe commented:
> Maybe the transaction commands need to return some kind of `TransactionIO[A]` monad, and we have a `TransactionIO` ~> `F` in the client?

TBH I think that would be over-engineering... 

> I'm not understanding why there inherently needs to be concurrency

It would be probably clearer if you try to do this in an imperative way (the way the Java library does) and then translate it to the FP world. 

```bash
> cmd.multi // Future[Unit]
> cmd.set("key1", "foo") // Future[Unit]
> cmd.set("key2", "bar") // Future[Unit]
> cmd.exec // Future[Unit]
```

In the example above, every command is sent asynchronously but performing side-effects (Futures execute immediately). Now compare that to the FP example:

```bash
> cmd.multi.unsafeRunSync() // Unit
> cmd.set("key1", "foo").unsafeRunSync() // Unit (blocks forever)
> ... // never gonna be reached
> cmd.exec.unsafeRunSync() // Unit
```

So by forking the operation `cmd.set("key1", "foo").start` we can have the same functionality as with the imperative `Future[Unit]`.

Daenyth commented:
Aha, I see - I didn't know that the java api returned `Future[Unit]` and expects you to fork them yourself. 

I presume this also hangs then?
```scala
for {
  _ <- cmd.multi // Future[Unit]
  _ <- cmd.set("key1", "foo") // Future[Unit]
  _ <- cmd.set("key2", "bar") // Future[Unit]
  _ <- cmd.exec // Future[Unit]
} yield ()
```

gvolpe commented:
Right, that hangs too because it's a sequence of actions depending on previous results (what `flatMap` is for). I haven't tried it though, but I'm almost sure. Will give it a try this evening when I get home.

I've been thinking overnight that `Lettuce` also exposes a synchronous API that will work for transactions without forking every single command since Redis replies immediately with an acknowledge that the command has been received. This will require to use `ContextShift` for blocking operations though. I'm not sure it's worth the amount of work (requires code duplication for the entire API).

I think having at least a simple `val tx = Transactional(cmd)` to run transactional commands is more than enough for an initial support for transactions. In the future this can be refined if we get better ideas.

gvolpe commented:
@Daenyth if no objections I'll merge this and publish a new milestone. In the issue tracker there are a few more tasks such as transactions in cluster environments. I'd like to have all these features and other transactions related issues / ideas before publishing a `0.8.0`.

Daenyth commented:
I think this is fine for an initial approach - a new api can be implemented later that gives more robust support.
